### PR TITLE
fix: “Generate constant” ignores the path prefix of the identifier

### LINKED ISSUE #12022: “Generate constant” ignores the path prefix of the identifier
Start with:

```rust
mod foo {}

fn bar() -> i32 {
    foo::A_CONSTANT
}
```

Run “Generate constant” on `A_CONSTANT`. The result is currently:

```rust
mod foo {}

const A_CONSTANT: i32 = ;
fn bar() -> i32 {
    foo::A_CONSTANT
}
```

Instead, the constant should be located in the named module:

```rust
mod foo {
    pub const A_CONSTANT: i32 = ;
}

fn bar() -> i32 {
    foo::A_CONSTANT
}
```

**rust-analyzer version**: 1c22537b3 2022-04-16 nightly and 24cf95762 2022-04-11 stable

**rustc version**: rustc 1.60.0 (7737e0b5c 2022-04-04)

### PR DESCRIPTION
**Pull request by harpsword**
fix #12022 

add abilities to generate constant with prefix path, even these mods in path are not exist.

## some examples

https://user-images.githubusercontent.com/10148927/168710096-59d5c594-0e4a-4ba7-bfb3-21c4b99409ea.mov

https://user-images.githubusercontent.com/10148927/168710111-cd6d3df4-58eb-4358-ae9e-791bfc408efa.mov

https://user-images.githubusercontent.com/10148927/168710115-601923b7-2164-4b9a-85a9-fbb4b29796a1.mov

https://user-images.githubusercontent.com/10148927/168710118-3d4bc9c1-758b-4e6f-9709-568fa920937d.mov

### CODE REVIEW

File: crates/ide-assists/src/handlers/generate_constant.rs [hunk 1]
```diff
@@ -47,20 +41,106 @@ pub(crate) fn generate_constant(acc: &mut Assists, ctx: &AssistContext) -> Optio
...
+                current_module = m;
+            }
+            Some(_) => {
+                return None;
+            }
+            None => {
+                not_exist_name_ref.push(name_ref_value);
+            }
+        }
+    }
```
Comment (Veykril) [inline crates/ide-assists/src/handlers/generate_constant.rs:75]: I guess this is fine for now, but ideally we want a helper function here that just creates/resolves a path for all its modules, as this here doesn't cover some cases properly.

File: crates/ide-assists/src/handlers/generate_constant.rs [hunk 2]
```diff
@@ -47,20 +41,106 @@ pub(crate) fn generate_constant(acc: &mut Assists, ctx: &AssistContext) -> Optio
...
         "Generate constant",
         target,
         |builder| {
-            builder.insert(
-                statement.syntax().text_range().start(),
-                format!("const {}: {} = $0;\n{}", constant_token, type_name, indent),
-            );
+            if file_id.is_some() {
+                builder.edit_file(file_id.unwrap());
+            }
```
Comment (Veykril) [inline crates/ide-assists/src/handlers/generate_constant.rs:92]:
```suggestion
            if let Some(file_id) = file_id {
                builder.edit_file(file_id);
            }
```

**Review by Veykril — COMMENTED** [review]
Sorry this took me so long to get back to, I have lost sight of this PR because it does some things we probably want proper support for (like creating inline modules for a path) and I was thinking about how to do that properly.

With that said, I am still not really sure how to do this nicely so I think its better to merge this now instead of keeping it blocked. So if you could fix the one `unwrap` usage, I'll go ahead and merge this.

File: crates/ide-assists/src/handlers/generate_constant.rs (hunk 2 above)
Comment (harpsword) [reply crates/ide-assists/src/handlers/generate_constant.rs]: done

**Comment by Veykril** [comment]
Thanks
@bors r+

**Pull request merged**